internal/generate/generate_modules.go

```diff
@@ -9,67 +10,57 @@ import (
 	"sync"
 	"text/template"
 
-	"github.com/commitdev/zero/internal/config"
+	"github.com/commitdev/zero/internal/config/projectconfig"
 	"github.com/commitdev/zero/internal/constants"
 	"github.com/commitdev/zero/internal/module"
 	"github.com/commitdev/zero/internal/util"
-	"github.com/commitdev/zero/pkg/util/exit"
 	"github.com/commitdev/zero/pkg/util/flog"
-	"github.com/k0kubun/pp"
 
 	"github.com/commitdev/zero/pkg/util/fs"
 )
 
-func GenerateModules(cfg *config.GeneratorConfig) {
-	var templateModules []*module.TemplateModule
+// Generate accepts a projectconfig struct and renders the templates for all referenced modules
+func Generate(projectConfig projectconfig.ZeroProjectConfig) error {
+	flog.Infof(":clock: Fetching Modules")
 
-	// TODO: Refactor this since the module struct is changing
-
-	// Initiate all the modules defined in the config
-	// for _, moduleConfig := range cfg.Modules {
-	//mod, err := module.NewTemplateModule(moduleConfig)
-
-	// if err != nil {
-	// 	exit.Error("module failed to load: %s", err)
-	// }
-	// templateModules = append(templateModules, mod)
-	// }
-
-	// Prompt for module params and execute each of the generator modules
-	for _, mod := range templateModules {
-		// TODO: read zero-project.yml instead
+	// Make sure module sources are on disk
+	wg := sync.WaitGroup{}
+	wg.Add(len(projectConfig.Modules))
+	for _, mod := range projectConfig.Modules {
+		go module.FetchModule(mod.Files.Source, &wg)
```

Comment:
Trying to understand how the go-getter stuff works
this will not download again if it's already on local?  for `Get()` on their [docs](https://godoc.org/github.com/hashicorp/go-getter#Get) they say
>"If dst already exists, Get will attempt to update it." 

and the `isLocal` function always receives the **remote-source** only, to do the `getter.Detect`, is getter.Get where the magic happens? 

https://github.com/commitdev/zero/blob/1a1ce930d6e87301004f09a58cb074164d24d245/internal/module/module.go#L23-L29

Reply:
Yeah right now it may fetch it again, but so far it's fast enough that it hasn't been a problem. I want to go over this stuff again at some point anyway to have these temporary files in a global location rather than written to a local `tmp` directory. At that point we can put more thought into if/how we want to cache these files.

Reply:
👍  I see, was just trying to figure out what it does

Reply:
Also the source location you provide can be local, not just remote, which is why we have that `isLocal` check there.